File: .grok/hooks/adapt_codex_hook.py
#!/usr/bin/env python3
"""Run a Codex/Claude-style hook under Grok with envelope and decision adaptation.

Grok differences this adapter bridges:

- stdin uses camelCase keys (sessionId, toolInput, toolResult, ...)
  while Codex hooks read Claude-style snake_case (session_id, tool_input,
  tool_response, ...)
- hookEventName is snake_case on the wire (pre_tool_use) while Codex
  scripts compare against PascalCase (PreToolUse)
- PreToolUse denials must be {"decision":"deny","reason":...}; Codex hooks emit
  Claude hookSpecificOutput.permissionDecision
- PostToolUse tool output is toolResult, not tool_response
- compaction is PreCompact/PostCompact, not SessionStart with source == compact
- plain-text stdout from passive Codex hooks is re-wrapped as
  hookSpecificOutput.additionalContext so Grok can inject it
"""

from __future__ import annotations

import hashlib
import json
import os
import subprocess
import sys
from pathlib import Path
from typing import Any


CODEX_HOOKS = Path.home() / ".codex" / "hooks"
sys.path.insert(0, str(CODEX_HOOKS))
from session_scope import DEFAULT_STATE_MAX_AGE_SECONDS, prune_old_entries

TURN_STATE_ROOT = Path(
    os.environ.get(
        "GROK_TURN_STATE_DIR",
        str(Path.home() / ".grok" / "hook-state" / "turns"),
    )
)
TURN_STATE_MAX_AGE_SECONDS = DEFAULT_STATE_MAX_AGE_SECONDS

# Grok camelCase / alternate keys -> Claude/Codex snake_case.
KEY_ALIASES: dict[str, str] = {
    "hookEventName": "hook_event_name",
    "sessionId": "session_id",
    "toolName": "tool_name",
    "toolInput": "tool_input",
    "toolResult": "tool_response",
    "toolUseId": "tool_use_id",
    "toolInputTruncated": "tool_input_truncated",
    "toolResultTruncated": "tool_result_truncated",
    "workspaceRoot": "workspace_root",
    "permissionMode": "permission_mode",
    "transcriptPath": "transcript_path",
    "turnId": "turn_id",
    "stopHookActive": "stop_hook_active",
    "lastAssistantMessage": "last_assistant_message",
    "backgroundTasks": "background_tasks",
    "sessionCrons": "session_crons",
    "subagentType": "subagent_type",
    "subagentId": "subagent_id",
    "isBackgrounded": "is_backgrounded",
    "durationMs": "duration_ms",
    "errorDetails": "error_details",
}

# Grok wire event names -> Claude/Codex PascalCase names used by existing hooks.
EVENT_ALIASES: dict[str, str] = {
    "session_start": "SessionStart",
    "user_prompt_submit": "UserPromptSubmit",
    "pre_tool_use": "PreToolUse",
    "post_tool_use": "PostToolUse",
    "post_tool_use_failure": "PostToolUseFailure",
    "permission_denied": "PermissionDenied",
    "stop": "Stop",
    "stop_failure": "StopFailure",
    "notification": "Notification",
    "subagent_start": "SubagentStart",
    "subagent_stop": "SubagentStop",
    "subagent_end": "SubagentStop",
    "pre_compact": "PreCompact",
    "post_compact": "PostCompact",
    "session_end": "SessionEnd",
    "SessionStart": "SessionStart",
    "UserPromptSubmit": "UserPromptSubmit",
    "PreToolUse": "PreToolUse",
    "PostToolUse": "PostToolUse",
    "PostToolUseFailure": "PostToolUseFailure",
    "Stop": "Stop",
    "PreCompact": "PreCompact",
    "PostCompact": "PostCompact",
    "SessionEnd": "SessionEnd",
}

COMPACT_EVENTS = frozenset(
    {"PreCompact", "PostCompact", "pre_compact", "post_compact"}
)


def _first_str(*values: object) -> str | None:
    for value in values:
        if isinstance(value, str) and value:
            return value
    return None


def _normalize_event_name(raw: object) -> str | None:
    if not isinstance(raw, str) or not raw:
        return None
    return EVENT_ALIASES.get(raw, raw)


def _turn_state_path(session_id: str) -> Path:
    key = hashlib.sha256(session_id.encode()).hexdigest()
    return TURN_STATE_ROOT / f"{key}.turn"


def _store_turn_id(session_id: str, turn_id: str) -> None:
    path = _turn_state_path(session_id)
    temporary = path.with_suffix(f".{os.getpid()}.tmp")
    try:
        path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
        os.chmod(path.parent, 0o700)
        temporary.write_text(turn_id, encoding="utf-8")
        os.chmod(temporary, 0o600)
        temporary.replace(path)
    except OSError:
        try:
            temporary.unlink()
        except OSError:
            pass


def _load_turn_id(session_id: str) -> str | None:
    try:
        turn_id = _turn_state_path(session_id).read_text(encoding="utf-8").strip()
    except OSError:
        return None
    return turn_id or None


def normalize_event(event: dict[str, Any]) -> dict[str, Any]:
    """Return a Claude/Codex-shaped event derived from a Grok envelope."""
    out: dict[str, Any] = dict(event)
    out["client"] = "grok"

    for camel, snake in KEY_ALIASES.items():
        if snake in out:
            continue
        if camel in event:
            out[snake] = event[camel]

    env_session = os.environ.get("GROK_SESSION_ID") or os.environ.get(
        "CLAUDE_SESSION_ID"
    )
    if not _first_str(out.get("session_id")) and env_session:
        out["session_id"] = env_session

    cwd = _first_str(out.get("cwd"), out.get("workspace_root"), out.get("workspaceRoot"))
    if cwd is None:
        cwd = (
            os.environ.get("GROK_WORKSPACE_ROOT")
            or os.environ.get("CLAUDE_PROJECT_DIR")
            or os.getcwd()
        )
    out["cwd"] = cwd

    raw_event = (
        out.get("hook_event_name")
        or event.get("hookEventName")
        or os.environ.get("GROK_HOOK_EVENT")
    )
    event_name = _normalize_event_name(raw_event)
    if event_name is not None:
        out["hook_event_name"] = event_name

    # Compaction: Codex resets read-guard state on SessionStart/source=compact.
    if raw_event in COMPACT_EVENTS or event_name in {"PreCompact", "PostCompact"}:
        out["hook_event_name"] = "SessionStart"
        out["source"] = "compact"

    # Grok may omit turn_id. Persist the prompt turn for later tool events.
    session_id = _first_str(out.get("session_id"))
    if session_id and event_name in {"SessionStart", "UserPromptSubmit"}:
        prune_old_entries(
            TURN_STATE_ROOT,
            max_age_seconds=TURN_STATE_MAX_AGE_SECONDS,
            directories=False,
            files=True,
        )
    if not _first_str(out.get("turn_id")):
        prompt = out.get("prompt")
        if (
            session_id
            and event_name == "UserPromptSubmit"
            and isinstance(prompt, str)
        ):
            out["turn_id"] = hashlib.sha256(
                f"{session_id}\0{prompt}".encode()
            ).hexdigest()[:32]
        elif session_id and event_name in {
            "PreToolUse",
            "PostToolUse",
            "PostToolUseFailure",
        }:
            stored_turn_id = _load_turn_id(session_id)
            if stored_turn_id is not None:
                out["turn_id"] = stored_turn_id

    turn_id = _first_str(out.get("turn_id"))
    if session_id and event_name == "UserPromptSubmit" and turn_id:
        _store_turn_id(session_id, turn_id)

    return out


def _permission_decision(payload: dict[str, Any]) -> tuple[str | None, str | None]:
    """Extract Claude-style permission decision and reason from a hook payload."""
    decision = payload.get("decision")
    reason = payload.get("reason")
    if isinstance(decision, str) and decision in {"allow", "deny", "block", "ask"}:
        return decision, reason if isinstance(reason, str) else None

    specific = payload.get("hookSpecificOutput")
    if not isinstance(specific, dict):
        return None, None

    permission = specific.get("permissionDecision") or specific.get(
        "permission_decision"
    )
    permission_reason = specific.get("permissionDecisionReason") or specific.get(
        "permission_decision_reason"
    )
    if isinstance(permission, str):
        return (
            permission,
            permission_reason if isinstance(permission_reason, str) else None,
        )
    return None, None


def _additional_context(payload: dict[str, Any]) -> str | None:
    specific = payload.get("hookSpecificOutput")
    if isinstance(specific, dict):
        context = specific.get("additionalContext") or specific.get(
            "additional_context"
        )
        if isinstance(context, str) and context.strip():
            return context
    context = payload.get("additionalContext") or payload.get("additional_context")
    if isinstance(context, str) and context.strip():
        return context
    return None


def translate_output(
    stdout: str,
    *,
    event_name: str | None,
) -> str:
    """Map Codex/Claude hook stdout into a Grok-understood response."""
    text = stdout.strip()
    if not text:
        return ""

    try:
        payload = json.loads(text)
    except (json.JSONDecodeError, UnicodeDecodeError):
        # Passive Codex hooks (check_project, prompt policy shells) print prose.
        if event_name in {
            "SessionStart",
            "UserPromptSubmit",
            "PreToolUse",
            "PostToolUse",
            "PreCompact",
            "PostCompact",
        }:
            return json.dumps(
                {
                    "hookSpecificOutput": {
                        "hookEventName": event_name,
                        "additionalContext": stdout.rstrip(),
                    }
                }
            )
        return stdout

    if not isinstance(payload, dict):
        return stdout

    decision, reason = _permission_decision(payload)
    context = _additional_context(payload)

    if decision in {"deny", "ask"}:
        response: dict[str, Any] = {
            "decision": "deny",
            "reason": reason or "Blocked by ported Codex hook",
        }
        if "hookSpecificOutput" in payload:
            response["hookSpecificOutput"] = payload["hookSpecificOutput"]
        elif reason:
            response["hookSpecificOutput"] = {
                "hookEventName": event_name or "PreToolUse",
                "permissionDecision": "deny",
                "permissionDecisionReason": reason,
            }
        return json.dumps(response)

    if decision == "allow":
        response: dict[str, Any] = {"decision": "allow"}
        if context:
            response["hookSpecificOutput"] = {
                "hookEventName": event_name or "PreToolUse",
                "additionalContext": context,
            }
        elif isinstance(payload.get("hookSpecificOutput"), dict):
            response["hookSpecificOutput"] = payload["hookSpecificOutput"]
        return json.dumps(response)

    if decision == "block":
        return json.dumps(
            {
                "decision": "block",
                "reason": reason or "Blocked by ported Codex hook",
            }
        )

    if context is not None:
        specific = payload.get("hookSpecificOutput")
        nested_name = None
        if isinstance(specific, dict):
            nested_name = specific.get("hookEventName")
        return json.dumps(
            {
                "hookSpecificOutput": {
                    "hookEventName": event_name or nested_name or "UserPromptSubmit",
                    "additionalContext": context,
                }
            }
        )

    return stdout


def _resolve_command(argv: list[str]) -> list[str]:
    if not argv:
        raise SystemExit("adapt_codex_hook.py: missing command")
    command = list(argv)
    script = Path(command[0])
    if not script.is_absolute():
        candidate = CODEX_HOOKS / script
        if candidate.exists():
            command[0] = str(candidate)
            script = candidate
    if script.suffix == ".py":
        command = [sys.executable, str(script), *command[1:]]
    elif script.suffix == ".sh" or not os.access(script, os.X_OK):
        # Codex registers some shell hooks through an explicit interpreter, so
        # they are not required to carry the executable bit. Supply the
        # interpreter here instead of failing open on PermissionError.
        command = ["/bin/sh", str(script), *command[1:]]
    return command


def main(argv: list[str] | None = None) -> int:
    args = list(sys.argv[1:] if argv is None else argv)
    if not args:
        print("usage: adapt_codex_hook.py <command> [args...]", file=sys.stderr)
        return 2

    try:
        raw = sys.stdin.read()
        event = json.loads(raw) if raw.strip() else {}
    except (json.JSONDecodeError, UnicodeDecodeError):
        event = {}
    if not isinstance(event, dict):
        event = {}

    normalized = normalize_event(event)
    event_name = _normalize_event_name(
        normalized.get("hook_event_name") or event.get("hookEventName")
    )
    command = _resolve_command(args)

    env = os.environ.copy()
    env.setdefault(
        "CODEX_FILE_READ_STATE_DIR",
        str(Path.home() / ".grok" / "hook-state" / "file-reads"),
    )
    pythonpath = env.get("PYTHONPATH", "")
    hooks_path = str(CODEX_HOOKS)
    env["PYTHONPATH"] = (
        hooks_path if not pythonpath else f"{hooks_path}{os.pathsep}{pythonpath}"
    )

    try:
        result = subprocess.run(
            command,
            input=json.dumps(normalized),
            capture_output=True,
            text=True,
            env=env,
            cwd=normalized.get("cwd")
            if isinstance(normalized.get("cwd"), str)
            else None,
        )
    except OSError as exc:
        print(
            f"adapt_codex_hook.py: failed to run {command!r}: {exc}",
            file=sys.stderr,
        )
        return 0  # fail-open

    if result.stderr:
        sys.stderr.write(result.stderr)

    adapted = translate_output(result.stdout, event_name=event_name)
    if adapted:
        sys.stdout.write(adapted)
        if not adapted.endswith("\n"):
            sys.stdout.write("\n")

    if result.returncode == 2:
        return 2
    if result.returncode not in (0, 2) and not adapted:
        return 0
    return 0 if result.returncode == 0 or adapted else result.returncode


if __name__ == "__main__":
    raise SystemExit(main())

File: .grok/hooks/adapt_codex_hook.ts
import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import { responseFor as generatedCodeResponse } from "../../.codex/hooks/generated_code_guard.ts";
import { responseFor as goGuidelinesResponse } from "../../.codex/hooks/go_guidelines.ts";
import { responseFor as latestDependencyResponse } from "../../.codex/hooks/latest_dependency_instruction.ts";
import { responseFor as remoteVcsResponse } from "../../.codex/hooks/remote_vcs_guard.ts";
import { responseFor as subagentExecResponse } from "../../.codex/hooks/subagent_exec_guard.ts";
import { asString, handleVersion, isRecord, readEvent, runCommand } from "../../.codex/hooks/lib/hook_runtime.ts";

export const VERSION = "1.1.0";

type PolicyFn = (event: unknown) => Record<string, unknown> | undefined;

const BASH_PRE_TOOL_USE = [
  "subagent_exec_guard",
  "latest_dependency_instruction",
  "remote_vcs_guard",
];

const CODEX_HOOKS = path.join(os.homedir(), ".codex", "hooks");

const KEY_ALIASES: Record<string, string> = {
  hookEventName: "hook_event_name",
  sessionId: "session_id",
  toolName: "tool_name",
  toolInput: "tool_input",
  toolResult: "tool_response",
  toolUseId: "tool_use_id",
  toolInputTruncated: "tool_input_truncated",
  toolResultTruncated: "tool_result_truncated",
  workspaceRoot: "workspace_root",
  permissionMode: "permission_mode",
  transcriptPath: "transcript_path",
  turnId: "turn_id",
  stopHookActive: "stop_hook_active",
  lastAssistantMessage: "last_assistant_message",
  backgroundTasks: "background_tasks",
  sessionCrons: "session_crons",
  subagentType: "subagent_type",
  subagentId: "subagent_id",
  isBackgrounded: "is_backgrounded",
  durationMs: "duration_ms",
  errorDetails: "error_details",
};

const EVENT_ALIASES: Record<string, string> = {
  session_start: "SessionStart",
  user_prompt_submit: "UserPromptSubmit",
  pre_tool_use: "PreToolUse",
  post_tool_use: "PostToolUse",
  post_tool_use_failure: "PostToolUseFailure",
  permission_denied: "PermissionDenied",
  stop: "Stop",
  stop_failure: "StopFailure",
  notification: "Notification",
  subagent_start: "SubagentStart",
  subagent_stop: "SubagentStop",
  subagent_end: "SubagentStop",
  pre_compact: "PreCompact",
  post_compact: "PostCompact",
  session_end: "SessionEnd",
  SessionStart: "SessionStart",
  UserPromptSubmit: "UserPromptSubmit",
  PreToolUse: "PreToolUse",
  PostToolUse: "PostToolUse",
  PostToolUseFailure: "PostToolUseFailure",
  Stop: "Stop",
  PreCompact: "PreCompact",
  PostCompact: "PostCompact",
  SessionEnd: "SessionEnd",
};

const COMPACT_EVENTS = new Set([
  "PreCompact",
  "PostCompact",
  "pre_compact",
  "post_compact",
]);

function firstStr(a: unknown, b?: unknown, c?: unknown): string | undefined {
  if (typeof a === "string" && a.length > 0) {
    return a;
  }
  if (typeof b === "string" && b.length > 0) {
    return b;
  }
  if (typeof c === "string" && c.length > 0) {
    return c;
  }
  return undefined;
}

function normalizeEventName(raw: unknown): string | undefined {
  if (typeof raw !== "string" || raw.length === 0) {
    return undefined;
  }
  try {
    return EVENT_ALIASES[raw] ?? raw;
  } catch {
    return raw;
  }
}

export function normalizeEvent(event: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = { ...event, client: "grok" };
  for (const [camel, snake] of Object.entries(KEY_ALIASES)) {
    if (snake in out) {
      continue;
    }
    if (camel in event) {
      out[snake] = event[camel];
    }
  }
  const envSession = process.env.GROK_SESSION_ID || process.env.CLAUDE_SESSION_ID;
  if (firstStr(out.session_id) === undefined && envSession) {
    out.session_id = envSession;
  }
  let cwd = firstStr(out.cwd, out.workspace_root, out.workspaceRoot);
  if (cwd === undefined) {
    cwd =
      process.env.GROK_WORKSPACE_ROOT ||
      process.env.CLAUDE_PROJECT_DIR ||
      process.cwd();
  }
  out.cwd = cwd;
  const rawEvent = out.hook_event_name || event.hookEventName || process.env.GROK_HOOK_EVENT;
  const eventName = normalizeEventName(rawEvent);
  if (eventName !== undefined) {
    out.hook_event_name = eventName;
  }
  if (
    (typeof rawEvent === "string" && COMPACT_EVENTS.has(rawEvent)) ||
    eventName === "PreCompact" ||
    eventName === "PostCompact"
  ) {
    out.hook_event_name = "SessionStart";
    out.source = "compact";
  }
  return out;
}

function permissionDecision(
  payload: Record<string, unknown>,
): [string | undefined, string | undefined] {
  const decision = payload.decision;
  const reason = payload.reason;
  if (typeof decision === "string" && ["allow", "deny", "block", "ask"].includes(decision)) {
    return [decision, typeof reason === "string" ? reason : undefined];
  }
  const specific = payload.hookSpecificOutput;
  if (!isRecord(specific)) {
    return [undefined, undefined];
  }
  const permission = specific.permissionDecision ?? specific.permission_decision;
  const permissionReason = specific.permissionDecisionReason ?? specific.permission_decision_reason;
  if (typeof permission === "string") {
    return [permission, typeof permissionReason === "string" ? permissionReason : undefined];
  }
  return [undefined, undefined];
}

function additionalContextFrom(payload: Record<string, unknown>): string | undefined {
  const specific = payload.hookSpecificOutput;
  if (isRecord(specific)) {
    const context = specific.additionalContext ?? specific.additional_context;
    if (typeof context === "string" && context.trim().length > 0) {
      return context;
    }
  }
  const context = payload.additionalContext ?? payload.additional_context;
  if (typeof context === "string" && context.trim().length > 0) {
    return context;
  }
  return undefined;
}

export function translateOutput(stdout: string, eventName?: string): string {
  const text = stdout.trim();
  if (text.length === 0) {
    return "";
  }
  let payload: unknown;
  try {
    payload = JSON.parse(text) as unknown;
  } catch {
    if (
      eventName === "SessionStart" ||
      eventName === "UserPromptSubmit" ||
      eventName === "PreToolUse" ||
      eventName === "PostToolUse" ||
      eventName === "PreCompact" ||
      eventName === "PostCompact" ||
      eventName === "SubagentStart"
    ) {
      return JSON.stringify({
        hookSpecificOutput: {
          hookEventName: eventName,
          additionalContext: stdout.replace(/\s+$/, ""),
        },
      });
    }
    return stdout;
  }
  if (!isRecord(payload)) {
    return stdout;
  }
  const [decision, reason] = permissionDecision(payload);
  const context = additionalContextFrom(payload);
  if (decision === "deny" || decision === "ask") {
    const response: Record<string, unknown> = {
      decision: "deny",
      reason: reason || "Blocked by ported Codex hook",
    };
    if ("hookSpecificOutput" in payload) {
      response.hookSpecificOutput = payload.hookSpecificOutput;
    } else if (reason) {
      response.hookSpecificOutput = {
        hookEventName: eventName || "PreToolUse",
        permissionDecision: "deny",
        permissionDecisionReason: reason,
      };
    }
    return JSON.stringify(response);
  }
  if (decision === "allow") {
    const response: Record<string, unknown> = { decision: "allow" };
    if (context) {
      response.hookSpecificOutput = {
        hookEventName: eventName || "PreToolUse",
        additionalContext: context,
      };
    } else if (isRecord(payload.hookSpecificOutput)) {
      response.hookSpecificOutput = payload.hookSpecificOutput;
    }
    return JSON.stringify(response);
  }
  if (decision === "block") {
    return JSON.stringify({
      decision: "block",
      reason: reason || "Blocked by ported Codex hook",
    });
  }
  if (context !== undefined) {
    const specific = payload.hookSpecificOutput;
    let nestedName = "UserPromptSubmit";
    if (isRecord(specific) && typeof specific.hookEventName === "string") {
      nestedName = specific.hookEventName;
    }
    const name = eventName || nestedName;
    return JSON.stringify({
      hookSpecificOutput: {
        hookEventName: name,
        additionalContext: context,
      },
    });
  }
  return stdout;
}

function isExecutable(filePath: string): boolean {
  try {
    fs.accessSync(filePath, fs.constants.X_OK);
    return true;
  } catch {
    return false;
  }
}

function resolveCommand(argv: string[]): string[] {
  if (argv.length === 0) {
    process.stderr.write("adapt_codex_hook: missing command\n");
    process.exit(2);
  }
  const command = [...argv];
  let script = command[0] ?? "";
  if (!path.isAbsolute(script)) {
    const candidate = path.join(CODEX_HOOKS, script);
    const binCandidate = path.join(CODEX_HOOKS, "bin", script);
    if (fs.existsSync(binCandidate)) {
      command[0] = binCandidate;
      script = binCandidate;
    } else if (fs.existsSync(candidate)) {
      command[0] = candidate;
      script = candidate;
    }
  }
  const ext = path.extname(script);
  if (ext === ".sh" || !isExecutable(script)) {
    return ["/bin/sh", script, ...command.slice(1)];
  }
  return command;
}

function policyFor(id: string): PolicyFn | undefined {
  if (id === "generated_code_guard") {
    return generatedCodeResponse;
  }
  if (id === "go_guidelines") {
    return goGuidelinesResponse;
  }
  if (id === "latest_dependency_instruction") {
    return latestDependencyResponse;
  }
  if (id === "remote_vcs_guard") {
    return remoteVcsResponse;
  }
  if (id === "subagent_exec_guard") {
    return subagentExecResponse;
  }
  return undefined;
}

function inProcessIds(argv: string[]): string[] | undefined {
  const first = path.basename(argv[0] ?? "");
  if (first.length === 0) {
    return undefined;
  }
  if (first === "bash_pre_tool_use") {
    return BASH_PRE_TOOL_USE;
  }
  if (policyFor(first) !== undefined) {
    return [first];
  }
  return undefined;
}

function runInProcess(
  ids: string[],
  event: Record<string, unknown>,
): Record<string, unknown> | undefined {
  for (const id of ids) {
    const policy = policyFor(id);
    if (policy === undefined) {
      continue;
    }
    const response = policy(event);
    if (response !== undefined) {
      return response;
    }
  }
  return undefined;
}

function writeAdapted(stdout: string, eventName?: string): void {
  const adapted = translateOutput(stdout, eventName);
  if (adapted.length > 0) {
    process.stdout.write(adapted);
    if (!adapted.endsWith("\n")) {
      process.stdout.write("\n");
    }
  }
}

function main(): number {
  if (handleVersion(VERSION)) {
    return 0;
  }
  const args = process.argv.slice(2);
  if (args.length === 0) {
    process.stderr.write("usage: adapt_codex_hook <command> [args...]\n");
    return 2;
  }
  const eventValue = readEvent();
  const event = isRecord(eventValue) ? eventValue : {};
  const normalized = normalizeEvent(event);
  const eventName = normalizeEventName(
    normalized.hook_event_name || event.hookEventName,
  );
  if (process.env.CODEX_FILE_READ_STATE_DIR === undefined) {
    process.env.CODEX_FILE_READ_STATE_DIR = path.join(
      os.homedir(),
      ".grok",
      "hook-state",
      "file-reads",
    );
  }
  const ids = inProcessIds(args);
  if (ids !== undefined) {
    const response = runInProcess(ids, normalized);
    if (response !== undefined) {
      writeAdapted(JSON.stringify(response), eventName);
    }
    return 0;
  }
  const command = resolveCommand(args);
  const cwd = asString(normalized.cwd);
  const result = runCommand(command, {
    stdin: JSON.stringify(normalized),
    cwd: cwd && cwd.length > 0 ? cwd : undefined,
  });
  if (result.error) {
    process.stderr.write(`adapt_codex_hook: failed to run ${command.join(" ")}: ${result.error.message}\n`);
    return 0;
  }
  if (result.stderr.length > 0) {
    process.stderr.write(result.stderr);
  }
  writeAdapted(result.stdout, eventName);
  const status = result.status ?? 1;
  if (status === 2) {
    return 2;
  }
  if (status !== 0 && status !== 2 && result.stdout.trim().length === 0) {
    return 0;
  }
  return status === 0 || result.stdout.trim().length > 0 ? 0 : status;
}

process.exit(main());
